Mathlib/Combinatorics/Nullstellensatz.lean

```diff
@@ -0,0 +1,307 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+import Mathlib.Algebra.MvPolynomial.Equiv
+import Mathlib.Algebra.Polynomial.Degree.Definitions
+import Mathlib.Data.Finsupp.MonomialOrder.DegLex
+import Mathlib.RingTheory.Ideal.Maps
+import Mathlib.RingTheory.MvPolynomial.Groebner
+import Mathlib.RingTheory.MvPolynomial.Homogeneous
+import Mathlib.RingTheory.MvPolynomial.MonomialOrder.DegLex
+
+/-! # Alon's Combinatorial Nullstellensatz
+
+This is a formalization of Noga Alon's Combinatorial Nullstellensatz. It follows [Alon_1999].
+
+We consider a family `S : σ → Finset R` of finite subsets of a domain `R`
+and a multivariate polynomial `f` in `MvPolynomial σ R`.
+The combinatorial Nullstellensatz gives combinatorial constraints for
+the vanishing of `f` at any `x : σ → R` such that `x s ∈ S s` for all `s`.
+
+- `MvPolynomial.eq_zero_of_eval_zero_at_prod_finset` :
+  if `f` vanishes at any such point and `f.degreeOf s < #(S s)` for all `s`,
+  then `f = 0`.
+
+- `combinatorial_nullstellensatz_exists_linearCombination`
+  If `f` vanishes at every such point, then it can be written as a linear combination
+  `f = linearCombination (MvPolynomial σ R) (fun i ↦ ∏ r ∈ S i, (X i - C r)) h`,
+  for some `h : σ →₀ MvPolynomial σ R` such that
+  `((∏ r ∈ S s, (X i - C r)) * h i).totalDegree ≤ f.totalDegree` for all `s`.
+
+- `combinatorial_nullstellensatz_exists_eval_nonzero`
+  a multi-index `t : σ →₀ ℕ` such that `t s < (S s).card` for all `s`,
+  `f.totalDegree = t.degree` and `f.coeff t ≠ 0`,
+  there exists a point `x : σ → R` such that `x s ∈ S s` for all `s` and `f.eval s ≠ 0`.
+
+## TODO
+
+- Applications
+- relation with Schwartz–Zippel lemma, as in [Rote_2023]
+
+## References
+
+- [Alon, *Combinatorial Nullstellensatz*][Alon_1999]
+
+- [Rote, *The Generalized Combinatorial Lason-Alon-Zippel-Schwartz
+  Nullstellensatz Lemma*][Rote_2023]
+
+-/
+
+open Finsupp
+
+open scoped Finset
+
+variable {R : Type*} [CommRing R]
+
+namespace MvPolynomial
+
+open Finsupp Function
+
+theorem _root_.Finsupp.option_embedding_add_single {σ : Type*}
+    {n : Option σ →₀ ℕ} {m : σ →₀ ℕ} {i : ℕ} :
+    (n = embDomain Embedding.some m + Finsupp.single none i) ↔
+      n none = i ∧ n.some = m := by
+  rw [Finsupp.ext_iff, Option.forall]
+  apply and_congr
+  · simp only [coe_add, Pi.add_apply, single_eq_same]
+    rw [embDomain_notin_range _ _ _ ?_, zero_add]
+    rintro ⟨s, hs⟩
+    exact Option.some_ne_none s hs
+  · rw [Finsupp.ext_iff]
+    apply forall_congr'
+    intro s
+    simp only [coe_add, Pi.add_apply, ne_eq, reduceCtorEq, not_false_eq_true, single_eq_of_ne,
+      add_zero, some_apply]
+    rw [← Embedding.some_apply, embDomain_apply, Embedding.some_apply]
```

Comment:
Can this move earlier?

Reply:
yes, to Mathlib/Data/Finsupp/Basic, and generalized from Nat to any AddCommMonoid.